### PR TITLE
Adding tf-keras based neural network to pipeline

### LINKED ISSUE #31: Add Neural Networks
**Is your feature request related to a problem? Please describe.**
We don't have Neural Networks in our machine learning workflow.

**Describe the solution you'd like**
Add Neural Networks.

**Additional context**
https://keras.io/getting_started/

### LINKED ISSUE #39: Add Data Scaling option/requirement
**Is your feature request related to a problem? Please describe.**
Some models, such as neural networks, need the data to be scaled.  Currently, there is normlized rdkit feature option, but early attempts seemed like it wasn't working well.  Additionally, it lacks control.  

**Describe the solution you'd like**
Implement and option (or requirement) to scale data using scikit's built in scaling functions, such as `StandardScalar()`.  Integrate it into the ml_model class and pipeline.  Store the type of scaling used for export  to knowledge graph.

### PR DESCRIPTION
**Pull request by Dlux804**
**Added Keras/Tensorflow dense neural network to the pipeline.** 
- Closes #31 
- Does not address Classification models.  
- Significant updates to fetching estimators (only regressors at this point).  
- Required updates to how parameters are passed to estimators (before and after tuning)
- Hyper-parameter tuning for neural network models with `skopt` 
- Keras models saved as `.h5` files for reloading
- Dropout added to `build_nn()` 
- All feature data is scaled via `sklearn StandardScaler()`.  Seems to have greatly improved predictions. (Closes #39 )

**Added output organization into zip folder**
- All output files, including pickles can be organized into a folder, which is then zipped.
- Option to delete the unzipped folder to avoid duplication

**Updated conda environment**
- Updated `mlapp.yml` with my most recent local environment
- Created a new file for storing the "pip install" packages that cannot be done with conda
- tensorflow is the gpu version

*Plenty of `#TODO` still left*

### CODE REVIEW

File: env_init/mlapp.yml [hunk 1]
```diff
@@ -1,37 +1,26 @@
...
   - mock
+  - tqdm
   - pip:
       - "git+git://github.com/bp-kelley/descriptastorus.git#egg=descriptastorus"
       - "git+https://github.com/swansonk14/chemprop.git"
       - "git+https://github.com/darenr/scikit-optimize.git#egg=scikit-optimize"
       - "git+https://github.com/Zsailer/pandas_flavor.git"
       - "git+https://github.com/mcs07/CIRpy.git"
 
+prefix: C:\Users\Adam\miniconda3\envs\mlapp
```
Comment (qle2) [inline env_init/mlapp.yml:25]: This won't affect CircleCi even if it's using your file path ?

File: main.py [hunk 2]
```diff
@@ -16,7 +17,8 @@ def main():
     print('ROOT Working Directory:', ROOT_DIR)
 
     # Asking the user to decide between running classification models or regression models
-    c = str(input("Enter c for classification, r for regression: "))
+    # c = str(input("Enter c for classification, r for regression: "))
+    c = 'r'
```
Comment (qle2) [inline main.py:21]: This was for quick testing correct?

File: main.py [hunk 3]
```diff
@@ -36,19 +38,22 @@ def main():
...
+        learner = ['ada', 'rf', 'svr', 'gdb', 'nn', 'knn']
+        learner = ['gdb', 'nn']
+
 
         # list of available featurization methods
-        feats = [[0], [0, 2], [0, 3], [0, 4], [0, 5], [2], [3], [4],
-                 [5]]  # Change this to change which featurizations are being tested (for regression)
+        feats = [[0], [0, 1], [0, 2], [0, 3], [0, 4], [0, 5], [1], [2], [3], [4],
+                 [5]]
+        feats = [[0]]#, [0, 2]]  # Change this to change which featurizations are being tested (for regression)
```
Comment (qle2) [inline main.py:48]: Duplicate variables. Comment one of these out

File: core/models.py [hunk 4]
```diff
@@ -57,31 +62,31 @@ def __init__(self, algorithm, dataset, target, feat_meth, tune=False, opt_iter=1
...
         """ Runs machine learning model. Stores results as class attributes."""
 
-        self.run_name = name.name(self.algorithm, self.dataset, self.feat_meth,
-                                  self.tuned)  # Create file nameprint(dict(vars(model1)).keys())
-        # TODO naming scheme currently must be called after featurization declared--adjust for robust
-
         if self.tuned:  # Do hyperparameter tuning
             self.make_grid()
-            self.hyperTune(n_jobs=6)
+            self.hyperTune(n_jobs=8)
```
[the remark this answers is not in the document]
Comment (andreshyer) [reply core/models.py:89]: I am not sure if that is the best way of doing it. I think it would more wise to have the user specify how many cores they are willing to give

File: core/grid.py [hunk 5]
```diff
@@ -90,12 +92,13 @@ def svr_paramgrid():
...
+    # with points predicted within a distance epsilon from the actual value
+    epsilon = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6]
 
     #  Kernel coefficient for 'rbf', 'poly' and 'sigmoid'.
 
-    gammas = [0.001, 0.005 ,0.01, 0.05 ,0.1, 0.5, 1]
+    gammas = [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1]
 
     # Degree of the polynomial kernel function ('poly')
     degrees = [1,2,3,4,5]
```
[the remark this answers is not in the document]
Comment (Dlux804) [reply core/grid.py:104]: No.  I envision that we will want to explore the SkLearn grid and random search for the sake of representing them in the graph.

File: core/models.py [hunk 6]
```diff
@@ -57,31 +62,31 @@ def __init__(self, algorithm, dataset, target, feat_meth, tune=False, opt_iter=1
...
+        # define run name used to save all outputs of model
+        self.run_name = name.name(self.algorithm, self.dataset, self.feat_meth, self.tuned)
 
         if not tune:  # if no tuning, no optimization iterations or CV folds.
             self.opt_iter = None
             self.cv_folds = None
-            self.regressor = self.regressor()  # make it callable to match Tune = True case
             self.tune_time = None
 
+    def reg(self):  # broke this out because input shape is needed for NN regressor to be defined.
```
[the remark this answers is not in the document]
Comment (Dlux804) [reply core/models.py:73]: Correct.  It should be `estimator()` or something like that. 
```suggestion
    def estimator(self):  # broke this out because input shape is needed for NN regressor to be defined.
```
^^ cool new feature I didn't know about "suggestion"

File: core/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Dlux804) [reply core/models.py:89]: This is not for GPU.  GPU must use n_jobs=1.  This is for CPU threads.  We could use `multiprocessing` to find the max and set that as default, but I want to keep manual override option.

File: core/storage.py [hunk 7]
```diff
@@ -59,23 +68,76 @@ def export_json(self):
...
+    path = os.getcwd()
+    base = self.run_name
+    zipdir = path+'\\'+base
+    print('Zipping output to {}'.format(base+'.zip'))
+    sleep(3)  # need to give time for files to move before zip.
+    shutil.make_archive(base, 'zip', zipdir)
+
+    if zip_only:  # remove unzipped folder to avoid duplicates
+        rmdir = 'rmdir ' + self.run_name + ' /s/q'
+        subprocess.Popen(rmdir, shell=True, stdout=subprocess.PIPE)  # run bash command
```
Comment (andreshyer) [inline core/storage.py:143]: This looks like this will likely break the unpickle function. This is a good add, but unpickle, and perhaps even pickle, might need to be modified to extract the pickle from the zipped folder.

File: env_init/mlapp.yml (hunk 1 above)
Comment (Dlux804) [reply env_init/mlapp.yml:25]: Not sure what exactly you are asking.  "Circle CI passed all tests" though.  So it was able to load the environment.

File: main.py (hunk 2 above)
Comment (Dlux804) [reply main.py:21]: Yes.  Also, I loath `input()` -- slows down things.

File: env_init/mlapp.yml (hunk 1 above)
Comment (andreshyer) [reply env_init/mlapp.yml:25]: Also, have you tested the environment to make sure that this environment works in our package?

File: core/storage.py [hunk 8]
```diff
@@ -39,18 +45,21 @@ def export_json(self):
     :param self:
     :return: writes json file
     """
+
```
Comment (andreshyer) [inline core/storage.py:48]: If this function is exporting csv and json files, the function name should likely be changed

File: main.py [hunk 9]
```diff
@@ -112,18 +122,49 @@ def example_model():
...
+    # Make predictions
+    predictions = model2.regressor.predict(self.test_features)
+
+    # Dataframe for replicate_model
+    pva = pd.DataFrame([], columns=['actual', 'predicted'])
+    pva['actual'] = model2.test_target
+    pva['predicted'] = predictions
+    r2 = r2_score(pva['actual'], pva['predicted'])
+    mse = mean_squared_error(pva['actual'], pva['predicted'])
+    rmme = np.sqrt(mean_squared_error(pva['actual'], pva['predicted']))
```
Comment (andreshyer) [inline main.py:164]: This is a nice function to have

File: main.py (hunk 2 above)
Comment (qle2) [reply main.py:21]: Should we add another input variable to choose estimator type? Something like: `model1 = models.MlModel(algorithm=alg, dataset=data, target=target, feat_meth=method, **estimator= "r"**,tune=True, cv=3, opt_iter=25)`
